Wikipedia/Code/MWKArticle.m

```diff
@@ -453,40 +404,75 @@ - (NSString*)debugDescription {
-    if (articleImageURL) {
-        [imageURLs addObject:articleImageURL];
-    }
+- (NSArray<NSURL*>*)imageURLsForSaving {
+    WMFImageTagList* tagList = [[[WMFImageTagParser alloc] init] imageTagListFromParsingHTMLString:self.articleHTML];
+    return [tagList imageURLsForSaving];
+}
 
+- (NSArray<MWKImage*>*)imagesForSaving {
+    return [[self imageURLsForSaving] bk_map:^id(NSURL* url){
+        return [[MWKImage alloc] initWithArticle:self sourceURL:url];
+    }];
+}
+
+-(NSArray<NSURL*>*)schemelessURLs:(NSArray<NSURL*>*)urls {
+    return [urls bk_map:^NSURL*(NSURL* url){
+        NSString* urlStr = url.absoluteString;
+        NSRange dividerRange = [urlStr rangeOfString:@"://"];
+        if (dividerRange.location != NSNotFound) {
+            return [NSURL URLWithString:[urlStr substringFromIndex:dividerRange.location + 1]];
+        }else{
+            return url;
+        }
+    }];
+}
+
+- (NSSet<NSURL*>*)allImageURLs {
+    WMFImageTagList* tagList = [[[WMFImageTagParser alloc] init] imageTagListFromParsingHTMLString:self.articleHTML];
+    
+    NSMutableSet<NSURL*>* imageURLs = [[NSMutableSet alloc] init];
+    //Note: use the 'imageURLsForGallery' and 'imageURLsForSaving' methods on WMFImageTagList so we don't have to parse twice.
+    [imageURLs addObjectsFromArray:[tagList imageURLsForGallery]];
+    [imageURLs addObjectsFromArray:[tagList imageURLsForSaving]];
+
+    NSArray<NSURL*>* lazilyFetchedHighResolutionGalleryImageURLs = [[self.dataStore imageInfoForTitle:self.title] valueForKey:WMF_SAFE_KEYPATH(MWKImageInfo.new, canonicalFileURL)];
+    [imageURLs addObjectsFromArray:[self schemelessURLs:lazilyFetchedHighResolutionGalleryImageURLs]];
+    
+    NSArray<NSURL*>* thumbURLs = [[self.dataStore imageInfoForTitle:self.title] valueForKey:WMF_SAFE_KEYPATH(MWKImageInfo.new, imageThumbURL)];
+    [imageURLs addObjectsFromArray:[self schemelessURLs:thumbURLs]];
+    
+    NSURL* articleImageURL = [NSURL wmf_optionalURLWithString:self.imageURL];
+    [imageURLs wmf_safeAddObject:articleImageURL];
+    
     NSURL* articleThumbnailURL = [NSURL wmf_optionalURLWithString:self.thumbnailURL];
-    if (articleThumbnailURL) {
-        [imageURLs addObject:articleThumbnailURL];
+    [imageURLs wmf_safeAddObject:articleThumbnailURL];
+
+    for (NSURL* url in imageURLs) {
+        NSAssert(url.scheme == nil, @"Non nil image url scheme detected! These must be nil here or we can get duplicate image urls in the set - ie the same url twice - once with the scheme and once without. This is because WMFImageTagParser returns urls without schemes because the html it parses does not have schemes - but the MWKImageInfo *does* have schemes in its 'canonicalFileURL' and 'imageThumbURL' properties because the api returns schemes.");
     }
 
     // remove any null objects inserted during above map/valueForKey operations
     return [imageURLs bk_reject:^BOOL (id obj) {
-        return [NSNull null] == obj;
+        return [obj isEqual:[NSNull null]];
```

Comment:
this can stay as == since [NSNull null] is a singleton - there could be a performance impact of calling obj's isEqual method, but probably not too bad since they're NSURLs